Feature/random letter list(wr9-56)

Kojinyoung7220 opened the pull request:
# Pull Request Template

## 📝 PR 설명

- 랜덤 편지 리스트 확인 로직 구현 (전체, 카테고리별)
- LetterRepository에 조회시 랜덤으로 가져오는 로직 구현
- 단위 테스트 작성

## 🔍 주요 변경사항

- [x] GET /api/random/{category} 엔드포인트 추가
- [x] 레파지토리에 findRandomLettersByCategory, findRandomLetters 메서드 구현

## 📸 스크린샷 (선택사항)

- UI 변경사항이 있다면 스크린샷을 첨부해주세요.

## 🧪 테스트

- [x] 단위 테스트 추가/수정
- [x] 테스트 커버리지 확인
- [x] 소나클라우드 품질 게이트 통과

## ✅ 체크리스트

- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 코드나 주석이 없나요?
- [x] 브랜치 네이밍 컨벤션을 따랐나요?
- [x] 관련 문서를 업데이트했나요?

## 🤝 관련 이슈

- closes #72

Kojinyoung7220 committed:
feat: 랜덤 편지 리스트 확인 기능 구현 (WR9-56)

- LetterMatchingController 구현
- LetterMatchingService에서 랜덤 편지 응답 로직 구현
- LetterRepository에서 랜덤 조회 쿼리 로직 구현
- RandomLetterResponse 랜덤편지 응답 dto 생성

Kojinyoung7220 committed:
Merge branch 'dev' into feature/random-letter-list(WR9-56)

Kojinyoung7220 committed:
feat: findRandomLetters 메서드에서 if문 조건 추가 (WR9-56)

Kojinyoung7220 committed:
test: LetterMatchingControllerTest, LetterMatchingServiceTest 단위 테스트 추가 (WR9-56)

leeys9423 commented on src/main/java/io/crops/warmletter/domain/letter/repository/LetterRepository.java:
```diff
@@ -10,4 +12,23 @@ public interface LetterRepository extends JpaRepository<Letter, Long> {
     List<Letter> findLettersByParentLetterId(Long parentLetterId);
 
-}
+    @Query(value =
+            "SELECT * " +
+            "FROM letters " +
+            "WHERE category = :category " +
+            "ORDER BY RAND() " +
+            "LIMIT :limit",
+            nativeQuery = true)
+    List<Letter> findRandomLettersByCategory(@Param("category") String category, @Param("limit") int limit);
+
+
+    @Query(value =
+            "SELECT * " +
+            "FROM letters " +
+            "ORDER BY RAND() " +
+            "LIMIT :limit",
+            nativeQuery = true)
+
+    List<Letter> findRandomLetters(@Param("limit") int limit);
```
1. 이 두 메서드 잘하면 하나로도 충분히 구현해내실 수 있을 것 같습니다!! (일단 제가 생각나는건 = 조회 대신 IN절을 사용하여 categories 리스트를 전달 하는방식)
2. nativeQuery대신 JPQL로 충분히 풀어낼수 있을 것 같아요.(nativeQuery는 특정 sql 에 종속되는 문법이고, 컴파일 시점에 문법오류를 체크할 수 없어 타입안정성이 떨어져 지양하는게 좋을 것 같아여)
3. 추후에는 고른 편지 노출을 위해 현재 프로세스보다 나은 방식을 생각해보시면 좋을 것 같아요!(중복된 편지가 조회될 수 있으므로)
    3-1. 가중치 (테이블에 viewCount를 넣어 제어해준다)
    3-2. 큐 기반 (편지리스트를 큐 기반으로 구현)
    3-3. 레디스 사용(Sorted Set 사용하면 될거라네요 이건 저도 검색한거라)

Kojinyoung7220 replied:
오 맞네요 in절이랑 Pageable쓰면 합칠 수 있고 서비스단 if문 분기도 줄일 수 있겠네요 
3번은 기능구현 다 마치구 고민 해보겠습니다 하하

leeys9423 commented on src/main/java/io/crops/warmletter/domain/letter/service/LetterMatchingService.java:
```diff
@@ -0,0 +1,52 @@
+    private final MemberRepository memberRepository;
+
+
+    public List<RandomLetterResponse> findRandomLetters(String category) {
```
String 타입의 category보다 Category라는 Enum값 사용하니까 타입 안정성을 위해서라도 Category enum을 활용하면 좋겠네요!

Kojinyoung7220 replied:
넵 습관처럼 스트링 나가버렸네용

leeys9423 commented on src/main/java/io/crops/warmletter/domain/letter/service/LetterMatchingService.java:
```diff
@@ -0,0 +1,52 @@
+
+        List<RandomLetterResponse> responses = new ArrayList<>();
+        for (Letter letter : letters) {
+            String zipCode = memberRepository.findById(letter.getWriterId())
```
한번 조회 시 member를 조회하는 쿼리가 5번씩 실행되게 되어있네요!!

letters 를 조회할 때 zipCode를 가져오는건 어떠세요?

비즈니스 로직상 편지를 쓴사람이 id를 안가질 순 없으니까 예외처리도 안해도 될 거 같기도 하고요!

Kojinyoung7220 replied:
지금 로직은 n+1 문제가 발생하겠네요 join으로 고치면 될 거 같습니다!

leeys9423 reviewed (commented):
고생하셨습니다

Kojinyoung7220 committed:
feat: 요구사항 추가, ID값 받아오는 로직 수정 (WR9-56)

- LetterResponse에 zipCode 필드 추가
- 기존 하드코딩된 id 값을 AuthFacade로 로그인한 유저 id 값 받는 로직으로 변경

Kojinyoung7220 committed:
test: 로직 변경으로 서비스단 mock 추가, 필드 검증 추가 (WR9-56)

- getPreviousLetters_success() 통합테스트 삭제

Kojinyoung7220 committed:
test: MailBoxService에 AuthFacade 사용으로 테스트 코드에 mock 추가 (WR9-56)

- todo list 주석 제거

Kojinyoung7220 committed:
feat: dto 생성자 추가, 컨트롤러 파라미터 타입 변경, 레파지토리 기능 수정 (WR9-56)

-  LetterMatchingService 서비스단 로직 수정
- LetterRepository 기존 2개로 나눠져 있던 쿼리 IN절과 페이징으로 변경
- findRandomLettersByCategory메서드에서 조인한 후 필요한 데이터를 한 번의 쿼리로 조회 -> n+1 문제 해결
- SELECT new 로 dto 빌더, 생성자 같이 사용

Kojinyoung7220 committed:
test: 기존 로직 요청 타입과 로직 변경으로 테스트 코드 수정 (WR9-56)

Kojinyoung7220 committed:
Merge branch 'dev' into feature/random-letter-list(WR9-56)
